model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt

```diff
@@ -236,16 +237,23 @@ class RepositoriesManager(val client: LocalModelClient) {
                 val version = CLVersion(versionHash, objectStore)
                 // Use a bulk query to make as few request to the underlying store as possible.
                 val bulkQuery = objectStore.newBulkQuery()
+                // It is unsatisfactory that we have to keep already emitted hashes in memory.
+                // But without changing the underlying model,
+                // we have to do this to not emit objects more than once.
+                val seenHashes = mutableSetOf<String>()
                 fun emitObjects(entry: KVEntryReference<*>) {
                     bulkQuery.get(entry).onSuccess {
                         channel.trySend(entry.getHash() to it!!.serialize())
                         for (referencedEntry in it!!.getReferencedEntries()) {
-                            emitObjects(referencedEntry)
+                            val wasSeenBefore = !seenHashes.add(referencedEntry.getHash())
+                            // Do not emit the object, if we already emitted id.
+                            if (!wasSeenBefore) {
+                                emitObjects(referencedEntry)
+                            }
                         }
                     }
                 }
-                channel.send(versionHash to kvStore.get(versionHash)!!)
-                emitObjects(version.treeHash!!)
+                emitObjects(KVEntryReference(versionHash, CPVersion.DESERIALIZER))
```

Comment:
@languitar This line is the main fix.

In the previous version, we only sent data for `treeHash`.
By using `KVEntryReference(versionHash, CPVersion.DESERIALIZER)` we send everything returned by `CLVersion.getReferencedEntries`.


```kt
    // CLVersion.getReferencedEntries
    override fun getReferencedEntries(): List<KVEntryReference<IKVValue>> {
        return listOfNotNull(
            treeHash,
            previousVersion,
            originalVersion,
            baseVersion,
            mergedVersion1,
            mergedVersion2,
            operationsHash,
        ) + (operations ?: arrayOf()).map { it.getReferencedEntries() }.flatten()
    }
```